### PR TITLE
tkkz - Don't allow assignments of a tutor to multiple course offerings with the same quarter

### PR DESCRIPTION
**Pull request by tkomarlu**

### CODE REVIEW

**Review by yinonrousso — CHANGES_REQUESTED** [review]
Want to double check with @scottpchow23. What .idea directory files should be uploaded to git? I have been getting conflicting information that the entire folder should be git ignored and info that says only specific files.

**Review by pconrad — CHANGES_REQUESTED** [review]
The .idea files should NOT be in this commit and pull request.

Can you put .idea in the .gitignore and then remove the .idea files from the commit?

**Review by pconrad — CHANGES_REQUESTED** [review]
The .iml file should be in the .gitignore as well.

It's very difficult to code review this with the extra files in the commit.

File: pom.xml [hunk 1]
```diff
@@ -49,12 +49,6 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-test</artifactId>
         </dependency>
```
Comment (pconrad) [inline pom.xml:51]: Was this a duplicate dependency?

It might be better to put the removal of this into a different pull request--that makes it easier to code review.

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java [hunk 2]
```diff
@@ -3,10 +3,7 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Controller;
 import org.springframework.ui.Model;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.*;
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java]: The Google Style guide says:

> Wildcard imports, static or otherwise, are not used.

https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java [hunk 3]
```diff
@@ -35,15 +34,31 @@ public TutorAssignmentController(TutorAssignmentRepository tutorAssignmentReposi
...
     Tutor tutor = tutorRepository.findById(tid)
         .orElseThrow(() -> new IllegalArgumentException("Invalid tutor Id:" + tid));
     CourseOffering courseOffering = courseOfferingRepository.findById(cid)
         .orElseThrow(() -> new IllegalArgumentException("Invalid course offering Id:" + cid));
-    TutorAssignment tutorAssignment = new TutorAssignment(tutor, courseOffering);
-    tutorAssignmentRepository.save(tutorAssignment);
-    model.addAttribute("tutorAssignments", tutorAssignmentRepository.findAll());
+    boolean isAssignedQuarter = false;
+
+    Iterator<TutorAssignment> iter = tutor.getTutorAssignments().iterator();
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java:60]: I wonder if this paragraph of code would be clearer if it were factored out into a method that has a name that matches what it does?

I stared at it a long time and was still not 100% sure I was clear on what it's supposed to be doing.  If were factored out into a method it would be easier to write unit tests for later as well.

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java [hunk 4]
```diff
@@ -35,15 +34,31 @@ public TutorAssignmentController(TutorAssignmentRepository tutorAssignmentReposi
...
+    while(iter.hasNext()){
+      if(courseOffering.getQuarter().equals(iter.next().getCourseOffering().getQuarter())){
+        isAssignedQuarter = true;
+      }
+    }
+
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java:61]: I wonder if it would be better to handle the error case first.   That's more typical, and let the main line code be the "good" case.

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/entities/CourseOffering.java [hunk 5]
```diff
@@ -78,4 +78,28 @@ public String toString() {
         return "CourseOffering{" + "id=" + id + ", course=" + course + ", quarter=" + quarter + ", instructor="
                 + instructor + '}';
     }
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs56/ucsb_courses_search/entities/CourseOffering.java:80]: Are duplicate courses in the same quarter (i.e. the same `course` name) permitted?

File: pom.xml (hunk 1 above)
Comment (tkomarlu) [reply pom.xml:51]: Okay fixed that.  Will make a new pull request for this

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java [hunk 6]
```diff
@@ -35,15 +34,31 @@ public TutorAssignmentController(TutorAssignmentRepository tutorAssignmentReposi
...
-    tutorAssignmentRepository.save(tutorAssignment);
-    model.addAttribute("tutorAssignments", tutorAssignmentRepository.findAll());
+    boolean isAssignedQuarter = false;
+
+    Iterator<TutorAssignment> iter = tutor.getTutorAssignments().iterator();
+    while(iter.hasNext()){
+      if(courseOffering.getQuarter().equals(iter.next().getCourseOffering().getQuarter())){
+        isAssignedQuarter = true;
+      }
+    }
```
Comment (scottpchow23) [inline src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java]: I think this actually is bugged because it won't check the first tutorAssignment in the list.

**Review by scottpchow23 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java (hunk 6 above)
Comment (tkomarlu) [reply src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java]: Fixed this by iterating through the iterable with a for loop

**Review by pconrad — APPROVED** [review]
There are some places this can be improved.  But in the interest of progress, I'll say LGTM.

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java (hunk 2 above)
Comment (tkomarlu) [reply src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java]: IntelliJ automatically creates wildcard imports rather than importing the specific class. Fixed imports to remove wildcard imports.

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java (hunk 3 above)
Comment (tkomarlu) [reply src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java:60]: This paragraph of code is meant to iterate through the Tutor Assignment repository and ensure that the course, given by the cid, is allowed to be assigned to a tutor. I refactored this code into a new method.

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java (hunk 4 above)
Comment (tkomarlu) [reply src/main/java/edu/ucsb/cs56/ucsb_courses_search/controllers/TutorAssignmentController.java:61]: Fixed code to reflect this

File: src/main/java/edu/ucsb/cs56/ucsb_courses_search/entities/CourseOffering.java (hunk 5 above)
Comment (tkomarlu) [reply src/main/java/edu/ucsb/cs56/ucsb_courses_search/entities/CourseOffering.java:80]: I believe there is nothing to verify that a course already exists. I will work on fixing this in a different branch

**Review by yinonrousso — COMMENTED** [review]
Looks good!

**Review by yinonrousso — APPROVED** [review]

**Comment by pconrad** [comment]
Looks like concerns in code review have been addressed.
